### PR TITLE
[versioning] EOS-26985: List object version

### PR DESCRIPTION
**Pull request by ajay-paratmandali**
Signed-off-by: Ajay Paratmandali <ajay.paratmandali@seagate.com>

# Problem Statement
- Optimize List object Version utility

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by ajay-paratmandali** [comment]
https://github.com/Seagate/cortx-s3server/pull/1649

Addressed comment in above PR

**Review by sanalpk — APPROVED** [review]

File: server/s3_list_object_versions_action.cc [hunk 1]
```diff
@@ -123,12 +105,21 @@ void S3ListObjectVersionsAction::validate_request() {
     max_keys = 1000;
   } else {
     int max_keys_temp = 0;
-    if ((!S3CommonUtilities::stoi(max_k, max_keys_temp)) ||
-        (max_keys_temp < 0)) {
+    if (!S3CommonUtilities::stoi(max_k, max_keys_temp)) {
       s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
              max_k.c_str());
       // TODO: Add invalid argument details to the response.
```
Comment (indrajitzagade) [inline server/s3_list_object_versions_action.cc:111]: Please remove this TODO now, same for TODO below as well

File: server/s3_list_object_versions_action.h [hunk 2]
```diff
@@ -35,17 +35,21 @@ class S3ListObjectVersionsAction : public S3BucketAction {
   std::shared_ptr<S3ObjectMetadataFactory> object_metadata_factory;
   std::shared_ptr<S3MotrKVSReader> motr_kv_reader;
   std::shared_ptr<MotrAPI> s3_motr_api;
-  bool include_marker_in_result;
-  bool check_any_keys_after_prefix;
+  bool include_marker_in_result = false;
```
Comment (indrajitzagade) [inline server/s3_list_object_versions_action.h:38]: Why is this changed to "false" now?

File: server/s3_list_object_versions_action.cc [hunk 3]
```diff
@@ -843,13 +839,10 @@ std::string& S3ListObjectVersionsAction::get_response_xml() {
...
-    free(encoded_str);
-  } else {
-    encoded_key_value = key_value;
+  if (encoding_type != "url") {
+    return key_value;
   }
-  return encoded_key_value;
+  const std::unique_ptr<char, void (*)(void*)> output(
+      evhttp_uriencode(key_value.c_str(), -1, 0), &free);
+  return output.get();
```
Comment (keithpine) [inline server/s3_list_object_versions_action.cc:847]: If `evhttp_uriencode` fails for whatever reason returns a NULL pointer, and you can't construct a std::string from NULL. I think this should be checking for NULL even though other code doesn't. 

```suggestion
  std::string encoded_key_value;
  const std::unique_ptr<char, void (*)(void*)> encoded_str(
      evhttp_uriencode(key_value.c_str(), -1, 0), &free);
  if (encoded_str) {
    encoded_key_value = encoded_str;
  }
  return encoded_key_value;
```

**Review by keithpine — CHANGES_REQUESTED** [review]

File: server/s3_list_object_versions_action.cc (hunk 1 above)
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: Removed.

File: server/s3_list_object_versions_action.h (hunk 2 above)
Comment (sjain09) [reply server/s3_list_object_versions_action.h]: changed to `true`

File: server/s3_list_object_versions_action.cc [hunk 4]
```diff
@@ -30,48 +30,30 @@
...
+      object_metadata_factory{
+          object_meta_factory ? std::move(object_meta_factory)
+                              : std::make_shared<S3ObjectMetadataFactory>()},
+      s3_motr_api{motr_api ? std::move(motr_api)
+                           : std::make_shared<ConcreteMotrAPI>()},
+      encoding_type{req->get_query_string_value("encoding-type")} {
+
+      bucket_metadata_factory = {
+          bucket_meta_factory ? std::move(bucket_meta_factory)
+                              : std::make_shared<S3BucketMetadataFactory>()};
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: `bucket_metadata_factory` is getting initialized in the parent class so removing this statement from here.

File: server/s3_list_object_versions_action.cc [hunk 5]
```diff
@@ -123,12 +105,21 @@ void S3ListObjectVersionsAction::validate_request() {
...
     int max_keys_temp = 0;
-    if ((!S3CommonUtilities::stoi(max_k, max_keys_temp)) ||
-        (max_keys_temp < 0)) {
+    if (!S3CommonUtilities::stoi(max_k, max_keys_temp)) {
       s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
              max_k.c_str());
       // TODO: Add invalid argument details to the response.
       set_s3_error("InvalidArgument");
+      set_s3_error_message("max-keys is not an integer");
+      set_invalid_argument("max-keys", max_k.c_str());
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: fixed.

File: server/s3_list_object_versions_action.cc [hunk 6]
```diff
@@ -123,12 +105,21 @@ void S3ListObjectVersionsAction::validate_request() {
...
+      send_response = true;
+    } else if (max_keys_temp < 0) {
+      s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
+             max_k.c_str());
+      // TODO: Add invalid argument details to the response.
+      set_s3_error("InvalidArgument");
+      set_s3_error_message("max-keys cannot be negative");
+      set_invalid_argument("max-keys", max_k.c_str());
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: Fixed coday issue.
Keeping `ArgumentValue` in the response body.

File: server/s3_list_object_versions_action.cc [hunk 7]
```diff
@@ -123,12 +105,21 @@ void S3ListObjectVersionsAction::validate_request() {
...
              max_k.c_str());
       // TODO: Add invalid argument details to the response.
       set_s3_error("InvalidArgument");
+      set_s3_error_message("max-keys is not an integer");
+      set_invalid_argument("max-keys", max_k.c_str());
+      send_response = true;
+    } else if (max_keys_temp < 0) {
+      s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
+             max_k.c_str());
+      // TODO: Add invalid argument details to the response.
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: removed

File: server/s3_list_object_versions_action.cc [hunk 8]
```diff
@@ -721,16 +711,16 @@ void S3ListObjectVersionsAction::add_object_version(const std::string& key,
...
   auto version = object_metadata_factory->create_object_metadata_obj(request);
-  if (version->from_json(value) != 0) {
-    json_error = true;
-    s3_log(S3_LOG_ERROR, request_id,
-           "Json Parsing failed. Index oid = "
-           "%" SCNx64 " : %" SCNx64 ", Key = %s, Value = %s\n",
-           object_version_list_index_oid.u_hi,
-           object_version_list_index_oid.u_lo, key.c_str(), value.c_str());
-  } else {
+  if (version->from_json(value) == 0) {
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: Updated.

File: server/s3_list_object_versions_action.cc [hunk 9]
```diff
@@ -123,12 +105,21 @@ void S3ListObjectVersionsAction::validate_request() {
...
   } else {
     int max_keys_temp = 0;
-    if ((!S3CommonUtilities::stoi(max_k, max_keys_temp)) ||
-        (max_keys_temp < 0)) {
+    if (!S3CommonUtilities::stoi(max_k, max_keys_temp)) {
       s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
              max_k.c_str());
       // TODO: Add invalid argument details to the response.
       set_s3_error("InvalidArgument");
+      set_s3_error_message("max-keys is not an integer");
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: Updated error message.

File: server/s3_list_object_versions_action.cc [hunk 10]
```diff
@@ -744,6 +734,12 @@ bool S3ListObjectVersionsAction::is_prefix_in_common_prefix(
 }
 
 std::string& S3ListObjectVersionsAction::get_response_xml() {
+  // Reserve capacity for the response xml based on versions_list length to
+  // avoid freqent memory re-location.
+  // reserve_cap = Max key length for object key is 1024 + few of hundred bytes
+  // for rest of the metadata.
+  auto reserve_cap = 1024 * (1 + versions_list.size());
```
[the remark this answers is not in the document]
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: Reverted the changes. Capped `max_keys` to 1000.

File: server/s3_list_object_versions_action.cc (hunk 3 above)
Comment (sjain09) [reply server/s3_list_object_versions_action.cc]: Updated.

**Review by indrajitzagade — APPROVED** [review]

**Comment by sjain09** [comment]
http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1674/ [Passed]
http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/674/ [Passed]

File: server/s3_list_object_versions_action.cc [hunk 11]
```diff
@@ -102,20 +100,26 @@ void S3ListObjectVersionsAction::validate_request() {
...
              max_k.c_str());
-      // TODO: Add invalid argument details to the response.
       set_s3_error("InvalidArgument");
       set_s3_error_message("max-keys cannot be negative");
-      set_invalid_argument("max-keys", max_k.c_str());
+      set_invalid_argument("max-keys", max_k);
       send_response = true;
     } else {
-      max_keys = max_keys_temp;
+      max_keys = max_keys_temp > kMaxKeys ? kMaxKeys : max_keys_temp;
```
Comment (keithpine) [inline server/s3_list_object_versions_action.cc:122]: For future reference, you can just use [std::min](https://en.cppreference.com/w/cpp/algorithm/min).

```suggestion
      max_keys = std::min(max_keys_temp, kMaxKeys);
```

**Review by keithpine — APPROVED** [review]
